tokenawarebandit.go

```diff
@@ -0,0 +1,614 @@
+	tokenRing *tokenRing
+	// partitioner used to map partition keys to tokens.
+	partitioner string
+}
+
+type tokenAwareBanditKeyspace struct {
+	// replicas allows to get replicas for token.
+	replicas tokenRingReplicas
+
+	// updating is true if we are currently updating the replica map for this keyspace.
+	updating bool
+
+	// needsUpdate is true if something changed while updating the replica map.
+	needsUpdate bool
+}
+
+// tokenAwareBanditHost keeps track of per-host state.
+type tokenAwareBanditHost struct {
+	// mu protects fields in tokenAwareBanditHost.
+	// If you need to hold both TokenAwareBanditHostPolicy.mu and tokenAwareBanditHost.mu,
+	// TokenAwareBanditHostPolicy.mu must be locked before tokenAwareBanditHost.mu.
+	mu sync.RWMutex
+	// latencyBuckets is a cyclic buffer storing history of latency measurements for a host.
+	// latencyBuckets[currentBucket] is the newest entry, the one to the right (with wrap) is the oldest entry.
+	latencyBuckets [bucketCount]latencyBucket
+	// currentBucket points to the newest bucket (the one we are writing to).
+	currentBucket int
+	// hostInfo of the host.
+	hostInfo *HostInfo
+}
+
+type latencyBucket struct {
+	sum   time.Duration
+	count int64
+}
+
+func (lb *latencyBucket) add(latency time.Duration) {
+	lb.sum += latency
+	lb.count += 1
+}
+
+// latency returns weighted average of latency measurements over time.
+// weight is linear from oldest bucket (1/bucketCount) to newest bucket (1.0).
+// h.mu must be held at least read-only.
+func (th *tokenAwareBanditHost) latency() (t time.Duration, ok bool) {
+	var sum time.Duration
+	var sumWeights int64
+	for i := 1; i <= len(th.latencyBuckets); i++ {
+		bucketIndex := (th.currentBucket + i) % bucketCount
+		if th.latencyBuckets[bucketIndex].count == 0 {
+			continue
+		}
+		sum += time.Duration(i) * th.latencyBuckets[bucketIndex].sum
+		sumWeights += int64(i) * th.latencyBuckets[bucketIndex].count
+	}
+	if sumWeights == 0 {
+		return 0, false
+	}
+	return sum / time.Duration(sumWeights), true
+}
+
+// recordLatency tracks the latency data point for this host.
+func (th *tokenAwareBanditHost) recordLatency(latency time.Duration) {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+	th.latencyBuckets[th.currentBucket].add(latency)
+}
+
+func (th *tokenAwareBanditHost) decay() {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+	th.currentBucket = (th.currentBucket + 1) % bucketCount
+	th.latencyBuckets[th.currentBucket] = latencyBucket{}
+}
+
+func (t *TokenAwareBanditHostPolicy) decay() {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	for _, host := range t.hosts {
+		host.decay()
+	}
+}
+
+func (t *TokenAwareBanditHostPolicy) decayLoop() {
+	ticker := time.NewTicker(t.decayPeriod / bucketCount)
+	for {
+		select {
+		case <-ticker.C:
+			t.decay()
+		case <-t.stopChan:
+			return
+		}
+	}
+}
+
+func (t *TokenAwareBanditHostPolicy) Init(s *Session) {
+	t.getKeyspaceMetadata = s.KeyspaceMetadata
+	go t.decayLoop()
+}
+
+func (t *TokenAwareBanditHostPolicy) Stop() {
```

Comment:
```suggestion
// Stop ...
// Call is blocking and executed only once after t.Init call. Is noop otherwise.
func (t *TokenAwareBanditHostPolicy) Stop() {
```

Let's mention that it's executed only once in docs.

Reply:
Added doc comment and updated Stop to be blocking (wait for goroutines to finish).

Reply:
Great! Thanks!